relay-server/src/processing/spans/mod.rs

```diff
@@ -366,27 +439,129 @@ pub struct Indexed;
+
+        for span in &self.spans {
+            if let Some(attachment) = &span.attachment {
+                attachment_quantity += attachment.body.len();
+                attachment_count += 1;
+            }
+        }
+        for attachment in &self.stand_alone_attachments {
+            attachment_quantity += attachment.body.len();
+            attachment_count += 1;
+        }
+
+        let mut quantities = smallvec::smallvec![];
+        if quantity > 0 {
+            quantities.push((DataCategory::Span, quantity));
+            quantities.push((DataCategory::SpanIndexed, quantity));
+        }
+        if attachment_quantity > 0 {
+            quantities.push((DataCategory::Attachment, attachment_quantity));
+            quantities.push((DataCategory::AttachmentItem, attachment_count));
+        }
+        quantities
     }
 }
 
 impl Counted for ExpandedSpans<Indexed> {
     fn quantities(&self) -> Quantities {
-        smallvec::smallvec![(DataCategory::SpanIndexed, self.spans.len())]
+        let mut attachment_quantity = 0;
+        let mut attachment_count = 0;
+
+        for span in &self.spans {
+            if let Some(attachment) = &span.attachment {
+                attachment_quantity += attachment.body.len();
+                attachment_count += 1;
+            }
+        }
+        for attachment in &self.stand_alone_attachments {
+            attachment_quantity += attachment.body.len();
+            attachment_count += 1;
+        }
+
+        let mut quantities = smallvec::smallvec![];
+
+        if !self.spans.is_empty() {
+            quantities.push((DataCategory::SpanIndexed, self.spans.len()));
+        }
+
+        if attachment_quantity > 0 {
+            quantities.push((DataCategory::Attachment, attachment_quantity));
+            quantities.push((DataCategory::AttachmentItem, attachment_count));
+        }
+
+        quantities
     }
 }
 
-impl CountRateLimited for Managed<ExpandedSpans<TotalAndIndexed>> {
+impl RateLimited for Managed<ExpandedSpans<TotalAndIndexed>> {
     type Error = Error;
+
+    async fn enforce<T>(
+        &mut self,
+        mut rate_limiter: T,
+        _: Context<'_>,
+    ) -> std::result::Result<(), Rejected<Self::Error>>
+    where
+        T: processing::RateLimiter,
+    {
+        let scoping = self.scoping();
+
+        for (category, quantity) in self.quantities() {
+            if matches!(category, DataCategory::Span | DataCategory::SpanIndexed) {
+                let limits = rate_limiter
+                    .try_consume(scoping.item(category), quantity)
+                    .await;
+
+                // If there is a span quota reject all the spans and the associated attachments.
+                if !limits.is_empty() {
+                    self.modify(|this, record_keeper| {
+                        record_keeper
+                            .reject_err(Error::from(limits), std::mem::take(&mut this.spans));
+                    });
```

Comment:
We can early-return here, so the attachments are also dropped.

```suggestion
             if !limits.is_empty() {
                let error = Error::from(limits);
                return Err(self.reject_err(error));
            }
```